### PR TITLE
Add converter method for ip adapters

### PR DESCRIPTION
**Pull request by fabiorigano**
# What does this PR do?

Fixes [5915 (comment)](https://github.com/huggingface/diffusers/pull/5915#discussion_r1413752931)

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten

### CODE REVIEW

**Comment by yiyixuxu** [comment]
can we maybe keep this function on the same file?

**Comment by fabiorigano** [comment]
> can we maybe keep this function on the same file?

done

I was wondering if it wouldn't be better to merge the two if blocks in `_load_ip_adapter_weights`, at line [713](https://github.com/huggingface/diffusers/blob/dfda680fea1fc186f79f6b2ec4b1fa559e37ae52/src/diffusers/loaders/unet.py#L713) and line [768](https://github.com/huggingface/diffusers/blob/dfda680fea1fc186f79f6b2ec4b1fa559e37ae52/src/diffusers/loaders/unet.py#L768)

File: src/diffusers/loaders/unet.py
Comment (yiyixuxu) [inline src/diffusers/loaders/unet.py:667]:
```suggestion
    def _convert_ip_adapter_image_proj_to_diffusers(self, state_dict):
```

File: src/diffusers/loaders/unet.py
Comment (yiyixuxu) [inline src/diffusers/loaders/unet.py:810]: can we have this function return the `image_projection` module instead?

File: src/diffusers/loaders/unet.py
Comment (sayakpaul) [reply src/diffusers/loaders/unet.py:667]: I wonder if we should move this method to https://github.com/huggingface/diffusers/blob/main/src/diffusers/loaders/ip_adapter.py. This way the UNet loader stays cleaner.

File: src/diffusers/loaders/unet.py
Comment (sayakpaul) [inline src/diffusers/loaders/unet.py:781]: Indeed better and cleaner!

File: src/diffusers/loaders/unet.py
Comment (sayakpaul) [inline src/diffusers/loaders/unet.py:780]:
```suggestion
            new_sd = self._convert_ip_adapter_image_proj_to_diffusers(image_proj_state_dict)
```

File: src/diffusers/loaders/unet.py
Comment (sayakpaul) [inline src/diffusers/loaders/unet.py:792]:
```suggestion
            new_sd = self._convert_ip_adapter_image_proj_to_diffusers(image_proj_state_dict)
```

**Review by sayakpaul — COMMENTED** [review]
Really nice work!

@yiyixuxu can we also run the SLOW tests related to IP adapters with these changes before merging?

**Review by patrickvonplaten — APPROVED** [review]
Ok to merge for me once @yiyixuxu is happy with it

**Comment by fabiorigano** [comment]
@yiyixuxu @sayakpaul @patrickvonplaten thank you for your time and digging into the code :)

**Pull request merged**
